File: cpswAsynPortDriverApp/deviceLibrarySrc/AxiXadc.cc
#include "AxiXadc.h"

#define addrSize 4

AxiXadc IAxiXadc::create(const char *name)
{
AxiXadcImpl v = CEntryImpl::create<AxiXadcImpl::element_type>(name);
Field f;
        f = IIntField::create("Temperature", 24, false, 4);
        v->CMMIODevImpl::addAtAddress( f , 0x200*addrSize);
        f = IIntField::create("VccInt", 24, false, 4);
        v->CMMIODevImpl::addAtAddress( f , 0x204*addrSize);
        f = IIntField::create("VccAux", 24, false, 4);
        v->CMMIODevImpl::addAtAddress( f , 0x208*addrSize);
        f = IIntField::create("VrefP", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x210*addrSize);
        f = IIntField::create("VrefN", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x214*addrSize);
        f = IIntField::create("VccBram", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x218*addrSize);
        f = IIntField::create("SupplyOffsetA", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x220*addrSize);
        f = IIntField::create("AdcOffsetA", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x224*addrSize);
        f = IIntField::create("AdcGainA", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x228*addrSize);
        f = IIntField::create("VccpInt", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x234*addrSize);
        f = IIntField::create("VccpAux", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x238*addrSize);
        f = IIntField::create("VccpDdr", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x23C*addrSize);
        f = IIntField::create("VinAux_0", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x240*addrSize);
        f = IIntField::create("VinAux01", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x244*addrSize);
        f = IIntField::create("VinAux02", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x248*addrSize);
        f = IIntField::create("VinAux03", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x24C*addrSize);
        f = IIntField::create("VinAux04", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x250*addrSize);
        f = IIntField::create("VinAux05", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x254*addrSize);
        f = IIntField::create("VinAux06", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x258*addrSize);
        f = IIntField::create("VinAux07", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x25C*addrSize);
        f = IIntField::create("VinAux08", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x260*addrSize);
        f = IIntField::create("VinAux09", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x264*addrSize);
        f = IIntField::create("VinAux10", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x268*addrSize);
        f = IIntField::create("VinAux11", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x26C*addrSize);
        f = IIntField::create("VinAux12", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x270*addrSize);
        f = IIntField::create("VinAux13", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x274*addrSize);
        f = IIntField::create("VinAux14", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x278*addrSize);
        f = IIntField::create("VinAux15", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x27C*addrSize);
        f = IIntField::create("TemperatureMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x280*addrSize);
        f = IIntField::create("VccIntMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x288*addrSize);
        f = IIntField::create("VccAuxMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x284*addrSize);
        f = IIntField::create("VccBramMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x28C*addrSize);
        f = IIntField::create("TemperatureMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x290*addrSize);
        f = IIntField::create("VccIntMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x294*addrSize);
        f = IIntField::create("VccAuxMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x298*addrSize);
        f = IIntField::create("VccBramMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x29C*addrSize);
        f = IIntField::create("VccpIntMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2A0*addrSize);
        f = IIntField::create("VccpAuxMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2A4*addrSize);
        f = IIntField::create("VccpDdrMax", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2A8*addrSize);
        f = IIntField::create("VccpIntMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2B0*addrSize);
        f = IIntField::create("VccpAuxMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2B4*addrSize);
        f = IIntField::create("VccpDdrMin", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2B8*addrSize);
        f = IIntField::create("Flag", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_JTGD", 1, false, 11);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_JTGR", 1, false, 10);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_REF", 1, false, 9);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccpDdr", 1, false, 7);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccpAux", 1, false, 6);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccpInt", 1, false, 5);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccBram", 1, false, 4);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccAux", 1, false, 2);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_VccInt", 1, false, 1);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        f = IIntField::create("Flag_Temperature", 1, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x2FC*addrSize);
        return v;
}

CAxiXadcImpl::CAxiXadcImpl(FKey key) : CMMIODevImpl(key, 0x00010000 >> 2, LE)
{
}

File: util/templates/AxiXadc.py
from __future__ import print_function
import os
from Cheetah.Template import Template

d = { 'name': "AxiXadc",
     'registers': [{'name'    : "Temperature",
                    'address' : "0x200",
                    'size'    : "24",
                    'lsBit'   : "4",
                    'type'    : "Status" },
                   {'name'    : "VccInt",
                    'address' : "0x204",
                    'size'    : "24",
                    'lsBit'   : "4",
                    'type'    : "Status" },
                   {'name'    : "VccAux",
                    'address' : "0x208",
                    'size'    : "24",
                    'lsBit'   : "4",
                    'type'    : "Status" },
                   {'name'    : "VrefP",
                    'address' : "0x210",
                    'type'    : "Status" },
                   {'name'    : "VrefN",
                    'address' : "0x214",
                    'type'    : "Status" },
                   {'name'    : "VccBram",
                    'address' : "0x218",
                    'type'    : "Status" },
                   {'name'    : "SupplyOffsetA",
                    'address' : "0x220",
                    'type'    : "Status" },
                   {'name'    : "AdcOffsetA",
                    'address' : "0x224",
                    'type'    : "Status" },
                   {'name'    : "AdcGainA",
                    'address' : "0x228",
                    'type'    : "Status" },
                   {'name'    : "VccpInt",
                    'address' : "0x234",
                    'type'    : "Status" },
                   {'name'    : "VccpAux",
                    'address' : "0x238",
                    'type'    : "Status" },
                   {'name'    : "VccpDdr",
                    'address' : "0x23C",
                    'type'    : "Status" },

#
                   {'name'    : "VinAux_0",
                    'address' : "0x240",
                    'type'    : "Status"},
                   {'name'    : "VinAux_1",
                    'address' : "0x244",
                    'type'    : "Status"},
                   {'name'    : "VinAux_2",
                    'address' : "0x248",
                    'type'    : "Status"},
                   {'name'    : "VinAux_3",
                    'address' : "0x24C",
                    'type'    : "Status"},
                   {'name'    : "VinAux_4",
                    'address' : "0x250",
                    'type'    : "Status"},
                   {'name'    : "VinAux_5",
                    'address' : "0x254",
                    'type'    : "Status"},
                   {'name'    : "VinAux_6",
                    'address' : "0x258",
                    'type'    : "Status"},
                   {'name'    : "VinAux_7",
                    'address' : "0x25C",
                    'type'    : "Status"},
                   {'name'    : "VinAux_8",
                    'address' : "0x260",
                    'type'    : "Status"},
                   {'name'    : "VinAux_9",
                    'address' : "0x264",
                    'type'    : "Status"},
                   {'name'    : "VinAux_10",
                    'address' : "0x268",
                    'type'    : "Status"},
                   {'name'    : "VinAux_11",
                    'address' : "0x26C",
                    'type'    : "Status"},
                   {'name'    : "VinAux_12",
                    'address' : "0x270",
                    'type'    : "Status"},
                   {'name'    : "VinAux_13",
                    'address' : "0x274",
                    'type'    : "Status"},
                   {'name'    : "VinAux_14",
                    'address' : "0x278",
                    'type'    : "Status"},
                   {'name'    : "VinAux_15",
                    'address' : "0x27C",
                    'type'    : "Status"},

#                   {'name'    : "halfSckPeriod",
#                    'address' : "0xA5",
#                    'type'    : "Status" },

                   {'name'    : "TemperatureMax",
                    'address' : "0x280",
                    'type'    : "Status" },
                   {'name'    : "VccIntMax",
                    'address' : "0x288",
                    'type'    : "Status" },
                   {'name'    : "VccAuxMax",
                    'address' : "0x284",
                    'type'    : "Status" },
                   {'name'    : "VccBramMax",
                    'address' : "0x28C",
                    'type'    : "Status" },
                   {'name'    : "TemperatureMin",
                    'address' : "0x290",
                    'type'    : "Status" },
                   {'name'    : "VccIntMin",
                    'address' : "0x294",
                    'type'    : "Status" },
                   {'name'    : "VccAuxMin",
                    'address' : "0x298",
                    'type'    : "Status" },
                   {'name'    : "VccBramMin",
                    'address' : "0x29C",
                    'type'    : "Status" },
                   {'name'    : "VccpIntMax",
                    'address' : "0x2A0",
                    'type'    : "Status" },
                   {'name'    : "VccpAuxMax",
                    'address' : "0x2A4",
                    'type'    : "Status" },
                   {'name'    : "VccpDdrMax",
                    'address' : "0x2A8",
                    'type'    : "Status" },
                   {'name'    : "VccpIntMin",
                    'address' : "0x2B0",
                    'type'    : "Status" },
                   {'name'    : "VccpAuxMin",
                    'address' : "0x2B4",
                    'type'    : "Status" },
                   {'name'    : "VccpDdrMin",
                    'address' : "0x2B8",
                    'type'    : "Status" },
                   {'name'    : "Flag",
                    'address' : "0x2FC",
                    'type'    : "Status" },
                   {'name'    : "Flag_JTGD",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "11",
                    'type'    : "Status" },
                   {'name'    : "Flag_JTGR",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "10",
                    'type'    : "Status" },
                   {'name'    : "Flag_REF",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "9",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccpDdr",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "7",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccpAux",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "6",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccpInt",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "5",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccBram",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "4",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccAux",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "2",
                    'type'    : "Status" },
                   {'name'    : "Flag_VccInt",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "1",
                    'type'    : "Status" },
                   {'name'    : "Flag_Temperature",
                    'address' : "0x2FC",
                    'size'    : "1",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "VccBramMin",
                    'address' : "0x29C",
                    'type'    : "Status" } ]

   }



os.mkdir('AxiXadc')
os.mkdir('AxiXadc/deviceLibrary')
os.mkdir('AxiXadc/AxiXadcSrc')
os.mkdir('AxiXadc/Db')
os.mkdir('AxiXadc/ui')

#device library
f=open('AxiXadc/deviceLibrary/AxiXadc.cc', 'w')
print(Template( file='builder.cc.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiXadc/deviceLibrary/AxiXadc.h', 'w')
print(Template( file='builder.h.tmpl', searchList = [d] ), file=f)
f.close()

#asyn driver
f=open('AxiXadc/AxiXadcSrc/Makefile', 'w')
print(Template( file='Makefile.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiXadc/AxiXadcSrc/AxiXadcDriver.cpp', 'w')
print(Template( file='templateDriver.cpp.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiXadc/AxiXadcSrc/AxiXadcDriver.h', 'w')
print(Template( file='templateDriver.h.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiXadc/AxiXadcSrc/AxiXadcDriverSupportInclude.dbd', 'w')
print(Template( file='templateDriverSupportInclude.dbd.tmpl', searchList = [d] ), file=f)
f.close()

#database substitutions
f=open('AxiXadc/Db/AxiXadc.substitutions', 'w')
print(Template( file='template.substitutions.tmpl', searchList = [d] ), file=f)
f.close()

#gui
f=open('AxiXadc/ui/AxiXadc.ui', 'w')
print(Template( file='templateGui.ui.tmpl', searchList = [d] ), file=f)
f.close()
